src/lib/coreaudio/CoreAudioDevice.cxx

```diff
@@ -0,0 +1,676 @@
+			stream_infos.push_back(info);
+		}
+		ret = true;
+	}
+	return ret;
+}
+
+void
+CoreAudioDevice::Close()
+{
+	if (!device_id)
+		return;
+
+	// Stop the device if it was started
+	Stop();
+
+	// Unregister the IOProc if available
+	RemoveIOProc();
+
+	// Make sure to release hog state
+	SetHogStatus(false);
+
+	if (buffer_size_restore)
+	{
+		SetBufferSize(buffer_size_restore);
+		buffer_size_restore = 0;
+	}
+	io_proc = nullptr;
+	device_id = 0;
+}
+
+void
+CoreAudioDevice::Start()
+{
+	if (!device_id || started)
+		return;
+
+	OSStatus err = AudioDeviceStart(device_id, io_proc);
+	if (err != noErr)
+		FormatError(macos_output_domain, "Unable to start device. Error = %s", GetError(err).c_str());
+	else
+		started = true;
+}
+
+void
+CoreAudioDevice::Stop()
+{
+	if (!device_id || !started)
+		return;
+
+	OSStatus err = AudioDeviceStop(device_id, io_proc);
+	if (err != noErr)
+		FormatError(macos_output_domain, "Unable to stop device. Error = %s", GetError(err).c_str());
+	started = false;
+	
+}
+
+bool
+CoreAudioDevice::AddIOProc(AudioDeviceIOProc callback_function, void* callback_data)
+{
+	assert(output_format.mFormatID != 0);
+	
+	// Allow only one IOProc at a time
+	if (!device_id || io_proc)
+		return false;
+
+	// Open the output stream and make the desired format active
+	output_stream.Open(stream_infos[output_stream_idx].stream_id);
+	if (output_format.mFormatFlags & kAudioFormatFlagIsNonMixable)
+		output_stream.SetVirtualFormat(&output_format);
+	output_stream.SetPhysicalFormat(&output_format);
+
+	// Create the callback
+	OSStatus err = AudioDeviceCreateIOProcID(device_id, callback_function, callback_data, &io_proc);
+	if (err != noErr)
+	{
+		FormatError(macos_output_domain, "Unable to add IOProc. Error = %s", GetError(err).c_str());
+		io_proc = nullptr;
+		return false;
+	}
+	// Start the device (callback)
+	Start();
+	return true;
+}
+
+bool
+CoreAudioDevice::RemoveIOProc()
+{
+	if (!device_id || !io_proc)
+		return false;
+	
+	Stop();
+	
+	// Close the output stream again and reverse format changes
+	output_stream.Close();
+
+	OSStatus err = AudioDeviceDestroyIOProcID(device_id, io_proc);
+	if (err != noErr)
+		FormatError(macos_output_domain, "Unable to remove IOProc. Error = %s", GetError(err).c_str());
+
+	io_proc = nullptr; // Clear the reference no matter what
+	
+	return true;
+}
+
+std::string
+CoreAudioDevice::GetName() const
+{
+	if (!device_id)
+		return "";
+
+	AudioObjectPropertyAddress  aopa;
+	aopa.mScope    = kAudioDevicePropertyScopeOutput;
+	aopa.mElement  = 0;
+	aopa.mSelector = kAudioDevicePropertyDeviceNameCFString;
+
+	std::string name;
+	CFStringRef deviceName = nullptr;
+	UInt32 propertySize = sizeof(deviceName);
+
+	OSStatus err = AudioObjectGetPropertyData(device_id, &aopa, 0, NULL, &propertySize, &deviceName);
+
+	if (err != noErr)
+	{
+		FormatError(macos_output_domain, "Unable to get device name - id: 0x%04x. Error = %s", (uint)device_id, GetError(err).c_str());
```

Comment:
Should this better throw an exception? I mean, what does it return now, an empty string? Do all callers check this condition?

Reply:
This method now throws an exception whenever no valid name can be returned